### ISSUE TITLE
Index kernel issue.

### ISSUE THREAD

**Issue by jikunshang** [intel/intel-extension-for-pytorch#505]
### Describe the bug

``` Python
import torch
import intel_extension_for_pytorch as ipex

ipex.version()

# probs = torch.ones((256,50272), dtype=torch.float32).to('cpu')
# indice= torch.range(0, 255,  dtype=torch.int32).to('cpu')
probs = torch.ones((256,50272), dtype=torch.float32).to('xpu')
indice = torch.range(0, 255,  dtype=torch.int32).to('xpu')

new = probs[indice]
print(new)
```
```
error message:
intel_extension_for_pytorch version:          2.1.10+xpu
intel_extension_for_pytorch git sha:          a12f9f650
submodule oneDNN sha:      887fb044ccd6308ed1780a3863c2c6f5772c94b3
submodule ideep sha:      ad79d0e1fa6b80db3c79a788896b9630d4e93940
/root/workspace/vllm-xpu/examples/ipex_test.py:10: UserWarning: torch.range is deprecated and will be removed in a future release because its behavior is inconsistent with Python's range builtin. Instead, use torch.arange, which produces values in [start, end).
  indice= torch.range(0, 255,  dtype=torch.int32).to('xpu')
/build/intel-pytorch-extension/csrc/gpu/aten/operators/Indexing.h:670: operator(): global id: [99520,0,0], local id: [192,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed
/build/intel-pytorch-extension/csrc/gpu/aten/operators/Indexing.h:670: operator(): global id: [99521,0,0], local id: [193,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed
/build/intel-pytorch-extension/csrc/gpu/aten/operators/Indexing.h:670: operator(): global id: [99522,0,0], local id: [194,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed
......
LIBXSMM_VERSION: main_stable-1.17-3651 (25693763)
LIBXSMM_TARGET: hsw [Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz]
Registry and code: 13 MB
Command: python3 examples/ipex_test.py
Uptime: 2.441513 s
[1]    2035021 IOT instruction (core dumped)  python3 examples/ipex_test.py
```

### Versions

Collecting environment information...
PyTorch version: 2.1.0a0+cxx11.abi
PyTorch CXX11 ABI: Yes
IPEX version: 2.1.10+xpu
IPEX commit: a12f9f650
Build type: Release

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: N/A
IGC version: 2024.0.0 (2024.0.0.20231017)
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.2.0-37-generic-x86_64-with-glibc2.35
Is XPU available: True
DPCPP runtime version: 2024.0
MKL version: 2024.0
GPU models and configuration: 
[0] _DeviceProperties(name='Intel(R) Arc(TM) A770 Graphics', platform_name='Intel(R) Level-Zero', dev_type='gpu, support_fp64=0, total_memory=15473MB, max_compute_units=512, gpu_eu_count=512)
Intel OpenCL ICD version: 23.30.26918.50-736~22.04
Level Zero version: 1.3.26918.50-736~22.04

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             44
On-line CPU(s) list:                0-43
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz
CPU family:                         6
Model:                              79
Thread(s) per core:                 2
Core(s) per socket:                 22
Socket(s):                          1
Stepping:                           1
CPU max MHz:                        3600.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           4399.89
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts md_clear flush_l1d
Virtualization:                     VT-x
L1d cache:                          704 KiB (22 instances)
L1i cache:                          704 KiB (22 instances)
L2 cache:                           5.5 MiB (22 instances)
L3 cache:                           55 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-43
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] intel-extension-for-pytorch==2.1.10+xpu
[pip3] numpy==1.26.2
[pip3] torch==2.1.0a0+cxx11.abi
[conda] N/A

**Comment by jgong5** [comment]
@zhuhaozhe can you take a look?

**Comment by zhuhaozhe** [comment]
hi, @jikunshang.
Thanks for reporting this.
I saw you code comment out `CPU` lines, does this issue only occured on `XPU` device?

**Comment by jikunshang** [comment]
> hi, @jikunshang. Thanks for reporting this. I saw you code comment out `CPU` lines, does this issue only occured on `XPU` device?

yes. CPU path works well. you can take a try:)

**Comment by jgong5** [comment]
Pass to @gujinghui since it is GPU specific.

**Comment by gujinghui** [comment]
@ashokei can you take a look, as it happens on ARC ? Thanks.

**Comment by jikunshang** [comment]
> @ashokei can you take a look, as it happens on ARC ? Thanks.

yes, I am testing on Arc A770.

**Comment by ashokei** [comment]
I just tested on PVC, and see the problem happening there too. So it is not ARC specific, will try to dig into sycl indexing kernel.

```
/build/intel-pytorch-extension/csrc/gpu/aten/operators/Indexing.h:670: operator(): global id: [12302,0,0], local id: [14,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed
/build/intel-pytorch-extension/csrc/gpu/aten/operators/Indexing.h:670: operator(): global id: [12303,0,0], local id: [15,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed
                                                                                                                                                                                                                
LIBXSMM_VERSION: main_stable-1.17-3651 (25693763)                                                                                                                                                               
LIBXSMM_TARGET: spr [Intel (R) Xeon (R) CPU Max 9470C]                                                                                                                                                          
Registry and code: 13 MB 
```

**Comment by ashokei** [comment]
indice dtype int64 works as a workaround for now.

```
import torch
import intel_extension_for_pytorch as ipex
probs = torch.ones((256,50272), dtype=torch.float32).to('xpu')
indice = torch.range(0, 255,  dtype=torch.int64).to('xpu')
new = probs[indice]
print(new)
```

**Comment by gujinghui** [comment]
@xiaolil1 pls take a look.

**Comment by xiaolil1** [comment]
We have reproduced this issue and filed JIRA to track it.
https://jira.devtools.intel.com/browse/PYTORCHDGQ-3883

**Comment by jikunshang** [comment]
> @jikunshang Suppose this issue has been fixed in [32dcfbd](https://github.com/intel/intel-extension-for-pytorch/commit/32dcfbda01d808567957c1d8c5b25e1765529c7f), could you please help to verify?

Thanks for your update! Will you provide a wheel package? It's quite complex to build from source code. And if you have verified the result, feel free to close this issue:)

**Comment by tye1** [comment]
@jikunshang  We don't provide wheel package for xpu-main branch. Let me close this now. Feel free to reopen if you still see this issue.
